source/common/websocket/codec.cc

```diff
@@ -0,0 +1,218 @@
+                                              : static_cast<uint8_t>(frame.payload_length_),
+                           output);
+  } else if (frame.payload_length_ <= 65535) {
+    // Set mask bit and 16-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xfe : 0x7e), output);
+    // Set 16-bit length
+    pushScalarToByteVector(htobe16(frame.payload_length_), output);
+  } else {
+    // Set mask bit and 64-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xff : 0x7f), output);
+    // Set 64-bit length
+    pushScalarToByteVector(htobe64(frame.payload_length_), output);
+  }
+  // Set masking key
+  if (frame.masking_key_) {
+    pushScalarToByteVector(htobe32(frame.masking_key_.value()), output);
+  }
+  return output;
+}
+
+bool Decoder::frameStart(uint8_t flags_and_opcode) {
+  // Validate opcode (last 4 bits)
+  uint8_t opcode = flags_and_opcode & 0x0f;
+  if (std::find(kFrameOpcodes.begin(), kFrameOpcodes.end(), opcode) != kFrameOpcodes.end()) {
+    frame_.opcode_ = opcode;
+    // Set final fragment flag
+    frame_.final_fragment_ = flags_and_opcode & 0x80;
+    return true;
+  }
+  decoding_error_ = true;
+  return false;
+}
+
+void Decoder::frameMaskFlag(uint8_t mask_and_length) {
+  // Set masked flag
+  if (mask_and_length & 0x80) {
+    masking_key_length_ = kMaskingKeyLength;
+  } else {
+    masking_key_length_ = 0;
+  }
+  // Set length (0 to 125) or length flag (126 or 127)
+  length_ = mask_and_length & 0x7F;
+}
+
+void Decoder::frameMaskingKey() {
+  frame_.masking_key_ = masking_key_;
+  masking_key_ = 0;
+}
+
+void Decoder::frameDataStart() {
+  frame_.payload_length_ = length_;
+  frame_.payload_ = std::make_unique<Buffer::OwnedImpl>();
+  if (length_ == 0) {
+    frameDataEnd();
+    state_ = State::KFrameHeaderFlagsAndOpcode;
+  } else {
+    state_ = State::KFramePayload;
+  }
+}
+
+void Decoder::frameData(const uint8_t* mem, uint64_t length) { frame_.payload_->add(mem, length); }
+
+void Decoder::frameDataEnd() {
+  if (!output_) {
+    output_ = std::vector<Frame>();
+  }
+  output_.value().push_back(std::move(frame_));
+  frame_.final_fragment_ = false;
+  frame_.opcode_ = 0;
+  frame_.payload_length_ = 0;
+  frame_.payload_ = nullptr;
+  frame_.masking_key_ = absl::nullopt;
+}
+
+void Decoder::doDecodeMaskFlagAndLength(const uint8_t c, const uint8_t*& mem,
+                                        uint64_t& slice_index) {
+  frameMaskFlag(c);
+  if (length_ == 0x7e) {
+    length_of_extended_length_ = kPayloadLength16Bit;
+    length_ = 0;
+    state_ = State::KFrameHeaderExtendedLength;
+  } else if (length_ == 0x7f) {
+    length_of_extended_length_ = kPayloadLength64Bit;
+    length_ = 0;
+    state_ = State::KFrameHeaderExtendedLength;
+  } else if (masking_key_length_ > 0) {
+    state_ = State::KFrameHeaderMaskingKey;
+  } else {
+    frameDataStart();
+  }
+  slice_index++;
+  mem++;
```

Comment:
It looks like these two argument are passed in simply to increment then. Instead, let's increment them at the caller.